qclib/gates/mcg.py

```diff
@@ -0,0 +1,382 @@
+        ancillae=controls[k_1:2 * k_1 - 2],
+        ctrl_state=ctrl_state_k_1
+    )
+    self.append(s_gate, [target])
+    self.mcx_v_chain_dirty(
+        controls=controls[k_1:],
+        target=target,
+        ancillae=controls[k_1 - k_2 + 1:k_1],
+        ctrl_state=ctrl_state_k_2
+    )
+    self.append(s_gate.inverse(), [target])
+    self.mcx_v_chain_dirty(
+        controls=controls[:k_1],
+        target=target,
+        ancillae=controls[k_1:2 * k_1 - 2],
+        ctrl_state=ctrl_state_k_1
+    )
+    self.append(s_gate, [target])
+    self.mcx_v_chain_dirty(
+        controls=controls[k_1:],
+        target=target,
+        ancillae=controls[k_1 - k_2 + 1:k_1],
+        ctrl_state=ctrl_state_k_2
+    )
+    self.append(s_gate.inverse(), [target])
+
+
+QuantumCircuit.linear_depth_mcv = linear_depth_mcv
+
+def linear_depth_any_mcsu2(
+    self,
+    unitary,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    ctrl_state: str=None
+):
+    """
+        Implements the gate decompostion of any gate in SU(2)
+        presented in Lemma 7.9 of https://arxiv.org/abs/quant-ph/9503016
+    """
+    theta, phi, lamb, _ = OneQubitEulerDecomposer._params_zyz(unitary)
+
+    a_gate, b_gate, c_gate = get_abc_operators(phi, theta, lamb)
+
+    self._apply_abc(
+        controls=controls,
+        target=target,
+        su2_gates=(a_gate, b_gate, c_gate),
+        ctrl_state=ctrl_state
+    )
+
+QuantumCircuit.linear_depth_any_mcsu2 = linear_depth_any_mcsu2
+
+def get_abc_operators(beta, gamma, delta):
+    """
+    Creates A,B and C matrices such that
+    ABC = I
+    """
+    # A
+    a_rz = RZGate(beta).to_matrix()
+    a_ry = RYGate(gamma / 2).to_matrix()
+    a_matrix = a_rz.dot(a_ry)
+
+    # B
+    b_ry = RYGate(-gamma / 2).to_matrix()
+    b_rz = RZGate(-(delta + beta) / 2).to_matrix()
+    b_matrix = b_ry.dot(b_rz)
+
+    # C
+    c_matrix = RZGate((delta - beta) / 2).to_matrix()
+
+    a_gate = UnitaryGate(a_matrix, label='A')
+    b_gate = UnitaryGate(b_matrix, label='B')
+    c_gate = UnitaryGate(c_matrix, label='C')
+
+    return a_gate, b_gate, c_gate
+
+
+def _apply_abc(
+    self,
+    controls: Union[QuantumRegister, List[Qubit]],
+    target: Qubit,
+    su2_gates: Tuple[UnitaryGate],
+    ctrl_state: str = None
+):
+    """
+        Applies ABC matrices to the quantum circuit according to theorem 5
+        of Iten et al. 2016 (arXiv:1501.06911).
+        Where su2_gates is a tuple of UnitaryGates in SU(2) eg.: (A,  B, C).
+    """
+    a_gate, b_gate, c_gate = su2_gates
+
+    if ctrl_state is not None:
+        for i, ctrl in enumerate(ctrl_state[::-1]):
+            if ctrl == '0':
+                self.x(controls[i])
+
+    if len(controls) < 3:
+
+        self.append(c_gate, [target])
+        self.mcx(controls, [target])
+        self.append(b_gate, [target])
+        self.mcx(controls, [target])
+        self.append(a_gate, [target])
+
+    else:
+        ancilla = controls[-1]
```

Comment:
@IsmaelCesar, could you check that everything is okay here?

Reply:
Everything seems okay.